### ISSUE TITLE
BUG? dropPropertyUniquenessConstraint throws strange wrapped exception

### ISSUE THREAD

**Issue by bHacklv** [neo4j/neo4j#4925]
Hi,

I've noticed you have strange/unnecessary difference in exception handling for these two methods:
**dropIndexDefinitions** and **dropPropertyUniquenessConstraint**.

When I try to do `ConstraintDefinition#drop()` on non-existing Constraint, behind the scenes it executes `dropPropertyUniquenessConstraint` and throws **ThisShouldNotHappenError**...

In same case for Index - `IndexDefinition#drop()` -> `dropIndexDefinitions` - throws **ConstraintViolationException**.

This is not consistent and should be fixed. Don't see why there should be two separate exceptions for Constraint and only one in Index.
Looks like it just stayed there as wip...

**Comment by FylmTM** [comment]
Hi,

More info on this issue.

Class: `org.neo4j.kernel.impl.coreapi.schema.PropertyUniqueConstraintDefinition`
Method: `org.neo4j.kernel.impl.coreapi.schema.PropertyUniqueConstraintDefinition#drop`

Code example: 

``` java
try (Transaction tx = db.beginTx()) {
    Schema schema = db.schema();
    ConstraintDefinition constraintDefinition = schema
            .constraintFor(TEST_CONSTRAINT_LABEL)
            .assertPropertyIsUnique(TEST_CONSTRAINT_PROP).create();
    constraintDefinition.drop();
    constraintDefinition.drop(); // <<<< there
    tx.success();
}
```

Running this code results in:

```
java.lang.Exception: Unexpected exception, expected<org.neo4j.graphdb.ConstraintViolationException> but was<org.neo4j.helpers.ThisShouldNotHappenError>
    at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:28)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at com.neueda.neo4j.driver.tests.integration.util.runner.IntegrationTestRunner.gentlyRunChild(IntegrationTestRunner.java:107)
    at com.neueda.neo4j.driver.tests.integration.util.runner.IntegrationTestRunner.runChild(IntegrationTestRunner.java:76)
    at com.neueda.neo4j.driver.tests.integration.util.runner.IntegrationTestRunner.runChild(IntegrationTestRunner.java:31)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
    at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
    at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: org.neo4j.helpers.ThisShouldNotHappenError: Developer: Mattias claims that: Unable to drop property unique constraint
    at org.neo4j.kernel.impl.coreapi.schema.SchemaImpl$GDBSchemaActions.dropPropertyUniquenessConstraint(SchemaImpl.java:475)
    at org.neo4j.kernel.impl.coreapi.schema.PropertyUniqueConstraintDefinition.drop(PropertyUniqueConstraintDefinition.java:45)
    at com.ericsson.neo4j.driver.tests.integration.schema.constraint.SchemaDropConstraintTest.testDropNotExistingIndexThrowConstraintViolationException(SchemaDropConstraintTest.java:97)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:19)
    ... 22 more
Caused by: org.neo4j.kernel.api.exceptions.schema.DropConstraintFailureException: Unable to drop constraint CONSTRAINT ON ( n:label[0] ) ASSERT n.property[0] IS UNIQUE: No such constraint CONSTRAINT ON ( n:label[0] ) ASSERT n.property[0] IS UNIQUE.
    at org.neo4j.kernel.impl.api.DataIntegrityValidatingStatementOperations.constraintDrop(DataIntegrityValidatingStatementOperations.java:161)
    at org.neo4j.kernel.impl.api.LockingStatementOperations.constraintDrop(LockingStatementOperations.java:303)
    at org.neo4j.kernel.impl.api.OperationsFacade.constraintDrop(OperationsFacade.java:734)
    at org.neo4j.kernel.impl.coreapi.schema.SchemaImpl$GDBSchemaActions.dropPropertyUniquenessConstraint(SchemaImpl.java:471)
    ... 33 more
Caused by: org.neo4j.kernel.api.exceptions.schema.NoSuchConstraintException: No such constraint CONSTRAINT ON ( n:label[0] ) ASSERT n.property[0] IS UNIQUE.
    at org.neo4j.kernel.impl.api.DataIntegrityValidatingStatementOperations.assertConstraintExists(DataIntegrityValidatingStatementOperations.java:232)
    at org.neo4j.kernel.impl.api.DataIntegrityValidatingStatementOperations.constraintDrop(DataIntegrityValidatingStatementOperations.java:156)
    ... 36 more
```

Interesting part here is:

```
Caused by: org.neo4j.helpers.ThisShouldNotHappenError: Developer: Mattias claims that: Unable to drop property unique constraint
...
Caused by: org.neo4j.kernel.api.exceptions.schema.DropConstraintFailureException: Unable to drop constraint CONSTRAINT ON ( n:label[0] ) ASSERT n.property[0] IS UNIQUE: No such constraint CONSTRAINT ON ( n:label[0] ) ASSERT n.property[0] IS UNIQUE.
```

Code that is responsible for exception handling is located:

``` java
// org/neo4j/kernel/impl/coreapi/schema/SchemaImpl.java:473
catch ( IllegalTokenNameException | TooManyLabelsException | DropConstraintFailureException e )
{
    throw new ThisShouldNotHappenError( "Mattias", "Unable to drop property unique constraint", e );
}
catch ( InvalidTransactionTypeKernelException e )
{
    throw new ConstraintViolationException( e.getMessage(), e );
}
```

We expected that we should receive `ConstraintViolationException` (similar to when index is deleted twice).

We are not sure what exception we should got when we delete same constraint twice in a row. But looks like that `ThisShouldNotHappenError` is definitely not the right one.

Thanks.

**Comment by chrisvest** [comment]
Closing since a fix has been merged and will be part of 2.3.
